drivers/sensor/MLX90640.zig

```diff
@@ -0,0 +1,922 @@
+//!
+//! Not all functionality from the datasheet is exposed in this driver
+//!
+//! The calculations were copied from https://github.com/netzbasteln/MLX90640-Thermocam
+//!
+const std = @import("std");
+const mdf = @import("../framework.zig");
+
+const Mlx90649Error = error{
+    BadPixels,
+};
+
+pub const MLX90640_Config = struct {
+    i2c: mdf.base.Datagram_Device,
+    clock: mdf.base.Clock_Device,
+    open_air_shift: f32 = 8,
+    emissivity: f32 = 0.95,
+};
+
+pub const MLX90640 = struct {
+    const Self = @This();
+
+    const control_register = packed struct {
+        enable_subpages_mode: u1,
+        reserved: u1,
+        enable_data_hold: u1,
+        enable_subpages_repeat: u1,
+        select_subpage: u3,
+        refresh_rate: u3,
+        resolution: u2,
+        reading_pattern: u1,
+        reserved2: u3,
+    };
+
+    const parameters = struct {
+        kVdd: i16 = 0,
+        vdd25: i16 = 0,
+        KvPTAT: f32 = 0,
+        KtPTAT: f32 = 0,
+        vPTAT25: i16 = 0,
+        alphaPTAT: f32 = 0,
+        gainEE: i16 = 0,
+        tgc: f32 = 0,
+        cpKv: f32 = 0,
+        cpKta: f32 = 0,
+        resolutionEE: u8 = 0,
+        calibrationModeEE: u8 = 0,
+        KsTa: f32 = 0,
+        ksTo: [5]f32 = undefined,
+        ct: [5]i16 = undefined,
+        alpha: [768]u16 = undefined,
+        alphaScale: u8 = 0,
+        offset: [768]i32 = undefined,
+        kta: [768]i8 = undefined,
+        ktaScale: u8 = 0,
+        kv: [768]i8 = undefined,
+        kvScale: u8 = 0,
+        cpAlpha: [2]f32 = undefined,
+        cpOffset: [2]i16 = undefined,
+        ilChessC: [3]f32 = undefined,
+        brokenPixels: [5]u16 = undefined,
+        outlierPixels: [5]u16 = undefined,
+    };
+
+    const registers = enum(u16) {
+        control = 0x800D,
+        status = 0x8000,
+        ram = 0x0400,
+        eeprom = 0x2400,
+        device_id = 0x2407,
+    };
+
+    eeprom: [832]u16 = undefined,
+    frame: [834]u16 = undefined,
+    frame_data: [834 * 2]u8 = undefined,
+    scratch_data: [768]f32 = undefined,
+    i2c: mdf.base.Datagram_Device,
+    clock_device: mdf.base.Clock_Device,
+    params: parameters,
+    emissivity: f32,
+    open_air_shift: f32,
+
+    scale_alpha: f32 = 0.000001,
+
+    frame_loop: [2]u1 = [2]u1{ 0, 1 },
+    refresh_rate_mask: u16 = 0b111 << 7,
+
+    pub fn init(cfg: MLX90640_Config) !Self {
+        return Self{
+            .i2c = cfg.i2c,
+            .clock_device = cfg.clock,
+            .open_air_shift = cfg.open_air_shift,
+            .emissivity = cfg.emissivity,
+            .params = parameters{},
+        };
+    }
+
+    fn write(self: *Self, reg: Self.registers, val: u16) !void {
+        const addr: u16 = @intFromEnum(reg);
+        var req = [4]u8{
+            @as(u8, @truncate(addr >> 8)),
+            @as(u8, @truncate(addr & 0xFF)),
+            @as(u8, @truncate(val >> 8)),
+            @as(u8, @truncate(val & 0xFF)),
+        };
```

Comment:
```suggestion
        var req = struct{addr_be: u16, val_be: u16}{
           .addr_be = std.mem.nativeToBig(addr),
           .var_be = std.mem.nativeToBig(val),
       };
```

Then `self.i2c_write(std.mem.asBytes(&req));` below